File: Algorithms/Java/Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedArrayToBST(int[] num) {
		// if (num.length == 0)
		// 	return null;
		return buildTree(num, 0, num.length - 1);
	}

	public TreeNode buildTree(int[] num, int a, int b) {
		if (a > b)
			return null;
		int mid = (a + b) / 2;
		TreeNode node = new TreeNode(num[mid]);
		node.left = buildTree(num, a, mid - 1);
		node.right = buildTree(num, mid + 1, b);
		return node;
	}
}

File: Algorithms/Python/Convert-Sorted-Array-to-BinarySearch-Tree.py
# Definition for a binary tree node.
# class TreeNode(object):

#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):

    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        def dfs(cur):
            if cur:
                mid = len(cur) // 2
                node = TreeNode(cur[mid])
                node.left = dfs(cur[:mid])
                node.right = dfs(cur[mid + 1:])
                return node
            else:
                return None

        return dfs(nums)

# METHOD 2


class Solution(object):

    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        def dfs(start, end):
            if start < end:
                mid = (start + end) // 2
                node = TreeNode(nums[mid])
                node.left = dfs(start, mid)
                node.right = dfs(mid + 1, end)
                return node
            else:
                return None

        return dfs(0, len(nums))
